### PR TITLE
rfcs: graph api: support safe softmax for sdpa

### PR DESCRIPTION
**Pull request by TaoLv**
This is to propose supporting safe softmax for sdpa in graph API.

A rendered version can be found at: [link](https://github.com/uxlfoundation/oneDNN/blob/lvtao/rfcs/safe-softmax/rfcs/20250321-graph-api-safe-softmax/README.md).

### CODE REVIEW

**Comment by LuFinch** [comment]
For option2, I think it is OK to make -inf coming from score as zero after safe softmax. Stock Pytorch does not distinguish whether -inf comes from score or mask, we could simply align with them. I prefer this option.

**Comment by TaoLv** [comment]
> For option2, I think it is OK to make -inf coming from score as zero after safe softmax. Stock Pytorch does not distinguish whether -inf comes from score or mask, we could simply align with them. I prefer this option.

Thank you for the comments, @LuFinch. I feel we need to understand if it's intended (or just simply overlooked) for not distinguishing the source of -inf in PyTorch. Do you have any reference to share?

**Comment by LuFinch** [comment]
> > For option2, I think it is OK to make -inf coming from score as zero after safe softmax. Stock Pytorch does not distinguish whether -inf comes from score or mask, we could simply align with them. I prefer this option.
> 
> Thank you for the comments, @LuFinch. I feel we need to understand if it's intended (or just simply overlooked) for not distinguishing the source of -inf in PyTorch. Do you have any reference to share?

Reference: https://github.com/pytorch/pytorch/blob/aa2ca7446dc39f4fb3daadd52f2db3fb69b5e831/aten/src/ATen/native/transformers/attention.cpp#L645

**Comment by TaoLv** [comment]
Hi @LuFinch , as we talked offline, we want to clarify the below questions:
1. Can we make the softmax in SDPA to do safe softmax unconditionally? Do you see any models or test cases in PyTorch that rely on the normal softmax for SDPA?
2. Can we make the softmax in all cases to do safe softmax unconditionally? That means not only for SDPA, but also for other models and test cases, softmax will return zeros for entire -inf row. We are asking this because we didn't see this request or behavior before this SDPA request.

If the answers are yes, we may don't need an API and just do safe softmax unconditionally in all kernels.

**Comment by LuFinch** [comment]
Hi @TaoLv,

For Question 1, I checked that stock Pytorch's FlashAtten CPU kernel and MemoryEfficientAttn GPU kernel already use safe softmax unconditionally. Hence, it is OK for make the softmax in SDPA to do safe softmax unconditionally. I don't see any models or tests in Pytorch rely on the normal softmax for SDPA.

For Question 2, I think we can't make the softmax in all cases to do safe softmax unconditionally because safe softmax is only for SDPA in stock Pytorch.  Doing this change might make softmax UT fail.

**Comment by TaoLv** [comment]
> Hi @TaoLv,
> 
> For Question 1, I checked that stock Pytorch's FlashAtten CPU kernel and MemoryEfficientAttn GPU kernel already use safe softmax unconditionally. Hence, it is OK for make the softmax in SDPA to do safe softmax unconditionally. I don't see any models or tests in Pytorch rely on the normal softmax for SDPA.
> 
> For Question 2, I think we can't make the softmax in all cases to do safe softmax unconditionally because safe softmax is only for SDPA in stock Pytorch. Doing this change might make softmax UT fail.

Thank you @LuFinch !
@mgouicem @dzarukin @petercad  Given these, do you think we can do safe softmax unconditionally in the backend without an API setting? If yes, I think I can update the RFC with the new option and close the PR #2979 .

**Comment by petercad** [comment]
@TaoLv Do you mind unconditional enabling of safe softmax for SDPA? That's fine by me.

**Comment by mgouicem** [comment]
@TaoLv IIUC, we cannot unconditionally do safe softmax.
From PyTorch perspective, they can hide internal use of safe_softmax since they expose an external [sdpa op](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html).
For non sdpa op, softmax preserves -inf (and I believe this is a necessary to propagate properly those for f16/f8 training where overflows need to be detected for rescaling).

From oneDNN perspective, we don't expose an sdpa op but pattern match it, and it seems awkward to document softmax op to have a different behavior based on how it is pattern matched.

File: rfcs/20250321-graph-api-safe-softmax/README.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+this option, we will not need to change anything in oneDNN.
+
+But the problem of option is also obvious:
+- There are more models use `-infinity` as masks. It's not possible to change
+  all of them to use other values. There is also an accuracy concern for it.
+- As PyTorch also changed to the new implementation, there is an inconsistency
+  between oneDNN and PyTorch which makes oneDNN integration fails in PyTorch
+  tests.
+
+### Proposal 2: Add an attribute `safe` to oneDNN softmax operation
```
[the remark this answers is not in the document]
Comment (TaoLv) [reply rfcs/20250321-graph-api-safe-softmax/README.md:54]: Thank you for the review. I updated both the RFC and the implementation PR.

**Comment by TaoLv** [comment]
Thank you for the feedback, @petercad and @mgouicem .

File: rfcs/20250321-graph-api-safe-softmax/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (dzarukin) [reply rfcs/20250321-graph-api-safe-softmax/README.md:54]: "Safe_final_v2_SWEAR"

**Review by dzarukin — COMMENTED** [review]
I've added a paragraph on primitive extension. @mgouicem - please take a look for a naming.

File: rfcs/20250321-graph-api-safe-softmax/README.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+typedef enum {
+    ...
+    /// Softmax
+    dnnl_softmax_accurate = 0x30000,
+    /// Logsoftmax
+    dnnl_softmax_log,
+    /// Softmax treating all infinity inputs as zero
+    dnnl_softmax_inf_as_zero,
+} dnnl_alg_kind_t;
+```
```
Comment (mgouicem) [inline rfcs/20250321-graph-api-safe-softmax/README.md:157]: I agree with this primitive extension for validation purpose.
Though I am not sure we want to expose it publicly as we don't have plan to optimize that, it is only for fallback and validation.
Can we keep this alg_kind internal, for use by Graph API and validation only?

If we expose it externally, we will have to very clearly document it that from accuracy perspective, it should not be used outside of sdpa pattern, and is public only for validation purpose.

Comment (dzarukin) [reply rfcs/20250321-graph-api-safe-softmax/README.md:162]: Removed the algorithm from public headers, updated an RFC, and all related codes.

**Comment by TaoLv** [comment]
The proposal has been implemented in oneDNN and released in v3.8.

**Review by rongzha1 — APPROVED** [review]

**Review by dzarukin — APPROVED** [review]
